src/v/serde/rw/variant.h

```diff
@@ -0,0 +1,99 @@
+/*
+ * Copyright 2024 Redpanda Data, Inc.
+ *
+ * Use of this software is governed by the Business Source License
+ * included in the file licenses/BSL.md
+ *
+ * As of the Change Date specified in that file, in accordance with
+ * the Business Source License, use of this software will be governed
+ * by the Apache License, Version 2.0
+ */
+
+#pragma once
+
+#include "base/vlog.h"
+#include "serde/rw/rw.h"
+
+#include <array>
+#include <utility>
+#include <variant>
+
+// # Variant Wire Compatibility:
+//
+// Variants are tricky w.r.t. compatibility, they are forward compatible if new
+// discrimiants are appended to the variant.
+//
+// Concretely that means that you can only read old versions of a variant if new
+// options are appended. Writing a new variant option should happen with a
+// feature barrier, so that an older reader will never read a newly written
+// variant.
+namespace serde {
+
+template<typename... T>
+void tag_invoke(tag_t<write_tag>, iobuf& out, std::variant<T...> v) {
+    write<size_t>(out, v.index());
+    std::visit(
+      [&out](auto&& v) {
+          write(out, std::forward<std::decay_t<decltype(v)>>(v));
+      },
+      std::move(v));
+}
+
+namespace detail {
+
+template<typename Variant>
+struct variant_discrimiant_factory {
+    using constructor = Variant (*)(iobuf_parser&, std::size_t);
+    using constructor_table
+      = std::array<constructor, std::variant_size_v<Variant>>;
+
+    template<size_t index>
+    static consteval void populate_table(constructor_table& lut) {
+        constructor reader = [](
+                               iobuf_parser& in, std::size_t bytes_left_limit) {
+            return Variant(
+              std::in_place_index<index>,
+              read_nested<std::variant_alternative_t<index, Variant>>(
+                in, bytes_left_limit));
+        };
+        lut[index] = reader;
+        if constexpr (index > 0) {
+            populate_table<index - 1>(lut);
+        }
+    }
```

Comment:
if you are looking for something without recursion, this (not tried yet) should work

```
constructor_table= []<size_t ...index> (std::index_sequence<index...>){
   return std::to_array<constructor>({  
       +([](iobuf_parser& in, size_t bytes_left_limit){
                  return Variant(
              std::in_place_index<index>,
              read_nested<std::variant_alternative_t<index, Variant>>(
                in, bytes_left_limit));
       })...
   
    });
}(std::make_index_sequence<std::variant_size_t<Variant>>());
```

Reply:
Wild I tried to come up with this. I had to use `std::to_array` but otherwise is 🔥

Reply:
yeah you need a way to force conversion of the lambdas to a common type.
here i used `+` to force decaying the lambda to a function pointer, https://gcc.godbolt.org/z/rGzdGPd4c
but i agree to_array<constructor> is a clearer for the reader